pkg/controller/metrics/metrics.go

```diff
@@ -743,13 +743,12 @@ func (ca *clusterAccumulator) setMetrics(total, installed, uninstalled, deprovis
 	}
 }
 
-// GetClusterDeploymentType returns the value of the hive.openshift.io/cluster-type label if set,
-// otherwise a default value.
-func GetClusterDeploymentType(obj metav1.Object) string {
-	if typeStr, ok := obj.GetLabels()[hivev1.HiveClusterTypeLabel]; ok {
+// GetLabelValue returns the value of the label if set, otherwise a default value.
+func GetLabelValue(obj metav1.Object, label string) string {
+	if typeStr, ok := obj.GetLabels()[label]; ok {
```

Comment:
Is it okay if the cd label's value is empty? If not:
```suggestion
	if typeStr := obj.GetLabels()[label]; typeStr != "" {
```
Note that this will treat "no such label" the same as "label exists and has empty string as its value".

Reply:
Right, so I do not remember when I discovered a problem - I distinctly remember us debating what a default value should be (like N/A or unknown). Nevertheless, I decided to track down the actual problem in the library because just 1 person remembering something like this is not ideal for future-proofing.
Okay, so `With` or `WithLabelValues` internally calls `GetMetricWith` from vec.go, and the problem comes when `getOrCreateMetricWithLabels` [creates](https://github.com/prometheus/client_golang/blob/254e5468413f19fb75cdad45f5ddc0b8c975188c/prometheus/vec.go#L514) the map with size equivalent to `labels` passed instead of reading from `metric.Desc` to determine the correct size (Note - even the `extractLabelValues` only relies on labels [provided](https://github.com/prometheus/client_golang/blob/254e5468413f19fb75cdad45f5ddc0b8c975188c/prometheus/vec.go#L616))

TL;DR
When the metric is created, it relies on the labels provided instead of metric definition as the dimension for creating the map. So if you provide empty string as label, it ignores it and creates the resultant metric map with less dimensions, consequently leading to always dropping some labels while observing.
Maybe `With` might not lead to such a problem - `WithLabelValues` definitely does. But if we know that happens, we don't want to treat them differently as a safety.

Reply:
Okay; so then my comment still applies. If the label _exists_ but happens to have an empty value, this code as written will cause it to be observed with an empty value rather than our default.

Now, based on the code you pointed to, it seems to me like that would be okay, because we would actually be passing in the empty string as an argument, as opposed to omitting the argument entirely, so the number of dimensions would still be correct. And if that's true, I would argue it's the appropriate thing to do: i.e. there's a difference in behavior for the user between the CR lacking the label ("unspecified") and the label existing with an empty value (the value _is_ specified, and it's `""`).

Whichever way we decide to do it, it should be consistent. Right now the fixed labels always convert `""` to the default, but the dynamic ones don't.